File: uniffi_bindgen/src/bindings/python/templates/UniffiTraitImpls.py
{# expects var `uniffi_trait_methods` #}
{%- if let Some(fmt) = uniffi_trait_methods.debug_fmt %}
{%-    let callable = fmt.callable %}
# The Rust `Debug::fmt` implementation.
def __repr__(self) -> {{ callable.return_type.type_name }}:
    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter -%}
{%- endif %}

{%- if let Some(fmt) = uniffi_trait_methods.display_fmt %}
{%-    let callable = fmt.callable %}
# The Rust `Display::fmt` implementation.
def __str__(self) -> {{ callable.return_type.type_name }}:
    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter -%}
{%- endif %}

{%- if let Some(eq) = uniffi_trait_methods.eq_eq %}
{%-    let callable = eq.callable %}
# The Rust `Eq::eq` implementation.
def __eq__(self, other: object) -> {{ callable.return_type.type_name }}:
    if not isinstance(other, {{ callable.self_type().unwrap().type_name }}):
        return NotImplemented

    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter -%}
{%- endif %}
{%- if let Some(ne) = uniffi_trait_methods.eq_ne %}
{%-    let callable = ne.callable %}
# The Rust `Eq::ne` implementation.
def __ne__(self, other: object) -> {{ callable.return_type.type_name }}:
    if not isinstance(other, {{ callable.self_type().unwrap().type_name }}):
        return NotImplemented
    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter -%}
{%- endif %}

{%- if let Some(hash) = uniffi_trait_methods.hash_hash %}
{%-    let callable = hash.callable %}
# The Rust `Hash::hash` implementation.
def __hash__(self) -> {{ callable.return_type.type_name }}:
    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter -%}

{%- endif %}

{%- if let Some(cmp) = uniffi_trait_methods.ord_cmp %}
{%-    let callable = cmp.callable %}
# The Rust `Ord::cmp` implementation.
# lol/sob, python3 got rid of the perfect python2 `.__cmp__()` :(
def __rust_cmp__(self, other) -> {{ callable.return_type.type_name }}:
    {% filter indent(4) -%}
    {% include "CallableBody.py" -%}
    {% endfilter %}

def __lt__(self, other) -> bool:
    return self.__rust_cmp__(other) < 0

def __le__(self, other) -> bool:
    return self.__rust_cmp__(other) <= 0

def __gt__(self, other) -> bool:
    return self.__rust_cmp__(other) > 0

def __ge__(self, other) -> bool:
    return self.__rust_cmp__(other) >= 0
{%- endif %}

File: uniffi_bindgen/src/bindings/ruby/templates/UniffiTraitImpls.rb
{# Template to generate standard Rust trait method implementations for a Ruby classes.
# Expects `trait_methods` to be a bound in the including template's scope.
# (e.g. `{%- let trait_methods = obj.uniffi_trait_methods() %}`)
#}

{%- if let Some(display_fmt) = trait_methods.display_fmt %}
# The Rust `Display::fmt` implementation.
def to_s
  result = ::{{ self.module_name() }}.rust_call(
    :{{ display_fmt.ffi_func().name() }},
    {{ display_fmt|lower_method_self_rb(self) }}
  )
  {{ self.lift_rb("result", display_fmt.return_type().unwrap())? }}
end
{%- endif %}

{%- if let Some(debug_fmt) = trait_methods.debug_fmt %}
# The Rust `Debug::fmt` implementation.
def inspect
  result = ::{{ self.module_name() }}.rust_call(
    :{{ debug_fmt.ffi_func().name() }},
    {{ debug_fmt|lower_method_self_rb(self) }}
  )
  {{ self.lift_rb("result", debug_fmt.return_type().unwrap())? }}
end
{%- endif %}

{%- if let Some(eq) = trait_methods.eq_eq %}
# The Rust `Eq::eq` implementation.
def ==(other)
  return false unless other.is_a?(self.class)
  result = ::{{ self.module_name() }}.rust_call(
    :{{ eq.ffi_func().name() }},
    {{ eq|lower_method_self_rb(self) }},
    {{ self.lower_rb("other", eq.arguments()[0].as_type().borrow())? }}
  )
  {{ self.lift_rb("result", eq.return_type().unwrap())? }}
end
{%- endif %}

{%- if let Some(hash) = trait_methods.hash_hash %}
# The Rust `Hash::hash` implementation.
def hash
  result = ::{{ self.module_name() }}.rust_call(
    :{{ hash.ffi_func().name() }},
    {{ hash|lower_method_self_rb(self) }}
  )
  {{ self.lift_rb("result", hash.return_type().unwrap())? }}
end

def eql?(other)
  self == other
end
{%- endif %}

{%- if let Some(cmp) = trait_methods.ord_cmp %}
# The Rust `Ord::cmp` implementation.
include Comparable

def <=>(other)
  # do we need this?
  # return nil unless other.is_a?(self.class)
  result = ::{{ self.module_name() }}.rust_call(
    :{{ cmp.ffi_func().name() }},
    {{ cmp|lower_method_self_rb(self) }},
    {{ self.lower_rb("other", cmp.arguments()[0].as_type().borrow())? }}
  )
  {{ self.lift_rb("result", cmp.return_type().unwrap())? }}
rescue
  nil
end
{%- endif %}
